lib/ansible/modules/unarchive.py

```diff
@@ -1144,30 +1173,14 @@ def main():
     # Run only if we found differences (idempotence) or diff was missing
     if res_args.get('diff', True) and not module.check_mode:
         # do we need to change perms?
-        top_folders = []
-        for filename in handler.files_in_archive:
+        for filename in handler.files_created_by_archive:
             file_args['path'] = os.path.join(b_dest, to_bytes(filename, errors='surrogate_or_strict'))
 
             try:
                 res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args['changed'], expand=False)
```

Comment:
`set_fs_attributes_if_different` handles OSError too, so I think this could be unindented further.

Do you think these changes should be included in the diff somehow? Maybe something like:

```python
            diff_attrs = {}
            res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args.get('changed', False), diff_attrs, expand=False)
            if diff_attrs:
                res_args.setdefault('diff', {})[filename] = diff_attrs
```